corehq/project_limits/gauge.py

```diff
@@ -0,0 +1,186 @@
+
+    def avg(self):
+        """
+        return average lag for a given topic
+        """
+        values = self.get_values()
+        return sum(values) / len(values) if values else None
+
+
+case_pillow_lag_gauge = Gauge(feature_key=CASE_SQL, scope_fn=get_all_kafka_partitons_for_topic)
+
+
+class GaugeLimiter:
+
+    def __init__(self, gauge, get_gauge_definition):
+        """
+        :param gauge: ``Gauge`` An instance of Gauge class
+        :param get_gauge_defintions: ``function``
+            A function to get config that GaugeLimiter will use to
+            figure out throttling conditions.
+        """
+        self.gauge = gauge
+        self.gauge_definition_fn = get_gauge_definition
+
+    def wait(self):
+        if not self.allow_usage():
+            time.sleep(self.gauge_definition.throttle_for_seconds)
+
+    def allow_usage(self):
+        if not self.gauge_definition:
+            return False
+        computed_value = None
+        if self.gauge_definition.actionable_metric == AVG:
+            computed_value = self.gauge.avg()
+        else:
+            computed_value = self.gauge.max()
+        if computed_value < self.gauge_definition.acceptable_value:
+            return True
+        return False
+
+    @property
+    def gauge_definition(self):
+        """
+        Should return an instace of subclass of ThrottleDefinition
+        """
+        return self.gauge_definition_fn(self.gauge.feature_key)
+
+
+class PillowLagGaugeLimiter(GaugeLimiter):
+
+    def _has_pillow_reported_recently(self):
+        time_window = 15  # time window in minutes in which throttling should be applied
+        time_before_window = datetime.now(tz=timezone.utc) - timedelta(minutes=time_window)
+        return self.gauge.get_last_reported_time() > time_before_window
+
+    def allow_usage(self):
+        if not self.gauge_definition:
+            # No throttling if config to throttle is not set
+            return True
+        if self._is_ideal_throttle_condition() and self._has_pillow_reported_recently():
+            return False
+        return True
+
+    def _is_ideal_throttle_condition(self):
+        """
+        The ideal throttle conditions are -
+        - If throttler is enabled in the config AND
+        - If we have reported values from scopes AND
+        - The follwing conditions are considered -
+            - If both max_value and average_value are set on the Defintion model,
+            then both should be breached in order to throttle.
+            - If only max_value is defined, then only max_value would be observed.
+            - If only average_value is set, then only average value would be observed.
+        """
+        if not self.gauge_definition.is_enabled:
+            # if throttling is disabled, don't throttle
+            return False
+
+        max_observed_value = self.gauge.max()
+        avg_observed_value = self.gauge.avg()
+
+        if not (max_observed_value and avg_observed_value):
```

Comment:
> Not sure if it could ever matter, but it seems safer to explicitly check for None here. What do you think?

That makes sense. I think we we ever set max value of some metric to 0 then this implementation will fail. 

>But this then brings up another question. Imagine a scenario where max_observed_value is not None and avg_observed_value is None and self.gauge_definition.max_value is not None and self.gauge_definition.average_value is None. What would be the correct thing to do in that case?

I don't think it is possible that `max_observed_value` is not None and `avg_observed_value` is None or vice versa because both `average` and `max` functions get values from `Gauge.get_values()` so either both of them will be None or none of them will be `None`. 
But I see your point of having race condition in fetching these values, I think in the case when one of the observed value is `None` then the following condition
```
if not (max_observed_value and avg_observed_value):
```
we will not throttle and will allow the usage and when the next batch will come then hopefully we will not see the race condition again and the throttling will work as expected. 

Does that answer your question or have I misunderstood your concern here?